include/circt/Support/FVInt.h

```diff
@@ -0,0 +1,592 @@
+    auto zeros = getZeroBits() | other.getZeroBits();
+    value &= other.value;
+    unknown |= other.unknown;
+    unknown &= ~zeros;
+    replaceZWithX();
+    return *this;
+  }
+
+  /// Compute the logical AND of this integer and a two-valued integer.
+  template <typename T>
+  FVInt &operator&=(T other) {
+    value &= other;
+    unknown &= other; // make 0 bits known
+    replaceZWithX();
+    return *this;
+  }
+
+  /// Compute the logical AND.
+  template <typename T>
+  FVInt operator&(const T &other) const {
+    auto v = *this;
+    v &= other;
+    return v;
+  }
+
+  /// Compute the logical OR of this integer and another. This implements the
+  /// following bit-wise truth table:
+  /// ```
+  ///     0 1 X Z
+  ///   +--------
+  /// 0 | 0 1 X X
+  /// 1 | 1 1 1 1
+  /// X | X 1 X X
+  /// Z | X 1 X X
+  /// ```
+  FVInt &operator|=(const FVInt &other) {
+    auto ones = getOneBits() | other.getOneBits();
+    value |= other.value;
+    unknown |= other.unknown;
+    unknown &= ~ones;
+    replaceZWithX();
+    return *this;
+  }
+
+  /// Compute the logical OR of this integer and a two-valued integer.
+  template <typename T>
+  FVInt &operator|=(T other) {
+    value |= other;
+    unknown &= ~other; // make 1 bits known
+    replaceZWithX();
+    return *this;
+  }
+
+  /// Compute the logical OR.
+  template <typename T>
+  FVInt operator|(const T &other) const {
+    auto v = *this;
+    v |= other;
+    return v;
+  }
+
+  /// Compute the logical XOR of this integer and another. This implements the
+  /// following bit-wise truth table:
+  /// ```
+  ///     0 1 X Z
+  ///   +--------
+  /// 0 | 0 1 X X
+  /// 1 | 1 0 X X
+  /// X | X X X X
+  /// Z | X X X X
+  /// ```
+  FVInt &operator^=(const FVInt &other) {
+    value ^= other.value;
+    unknown |= other.unknown;
+    replaceZWithX();
+    return *this;
+  }
+
+  /// Compute the logical XOR of this integer and a two-valued integer.
+  template <typename T>
+  FVInt &operator^=(const T &other) {
+    value ^= other;
+    replaceZWithX();
+    return *this;
+  }
+
+  /// Compute the logical XOR.
+  template <typename T>
+  FVInt operator^(const T &other) const {
+    auto v = *this;
+    v ^= other;
+    return v;
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Arithmetic Operators
+  //===--------------------------------------------------------------------===//
+
+  /// Compute the negation of this integer. If any bits are unknown, the entire
+  /// result is X.
+  void negate() {
+    value.negate();
+    setAllXIfAnyUnknown();
+  }
+
+  /// Compute the negation of this integer.
+  FVInt operator-() const {
+    auto v = *this;
+    v.negate();
+    return v;
+  }
+
+  /// Compute the addition of this integer and another. If any bits in either
+  /// integer are unknown, the entire result is X.
+  FVInt &operator+=(const FVInt &other) {
+    value += other.value;
+    setAllXIfAnyUnknown(other);
```

Comment:
nit: It might be a bit too conservative to mark everything unknown if lower bits are known, e.g. `x10 + x10 = x00`. Does IEEE 1800 define 4-value logic arithmetic as well as bitwise logical operation?

Reply:
Good point. I've found the following nugget in IEEE 1800-2017 "11.4.3 Arithmetic operators" hidden between tables:

> For the arithmetic operators, if any operand bit value is the unknown value x or the high-impedance value z, then the entire result value shall be x.

So I guess conservative is fine/desired?

Reply:
The "fun" really starts when you get to the unary arithmetic operators. What's the expected output here?
```
module test;
  logic [3:0] foo = 4'b01XZ;
  initial begin
    $display("None: %b", foo);
    $display("Plus: %b", +foo);
    $display("Minus: %b", -foo);
  end
endmodule
```
I've seen simulators disagree on the handling of the unary plus.

Reply:
Nasty! I guess from that line in the standard you'd expect to see `foo` unchanged, and both `+foo` and `-foo` to be all X since they count as arithmetic operators? Unless there's a line in the standard that says unary `+` can be treated as a no-op… But if that doesn't exist, we'd need a `moore.x_if_any_unknown` style op 😅

Reply:
Check out Table 11-5 right above the sentence you quoted.
![cantbeliveididthis](https://github.com/user-attachments/assets/5e19d020-aa70-462d-aa59-edde5ba0ad45)
I'll see myself out....

But most simulators seem to opt for "same as m".

Reply:
What a beautiful spec... But yeah, `+m = m` feels like the right thing to do intuitively, but it also seems to violate the spec. We can always drop an op that handles the X-ification for `+`.

Reply:
> So I guess conservative is fine/desired?

Oh, yeah. Thank you for pointing it out!